File: libs/core---esp32/crypto.cpp
#include "pxt.h"
#include "mbedtls/sha256.h"

// https://eprint.iacr.org/2012/156.pdf - in case more hash functions are needed

namespace crypto {

/*
 * Compute cryptographic SHA256 hash of the concatenation of buffers. Returns 32-byte buffer.
 */
//% promise
Buffer _sha256(RefCollection *buffers) {
    mbedtls_sha256_context sha256_ctx;
    uint8_t output[32];

    auto bufs = buffers->getData();
    auto len = buffers->length();

    mbedtls_sha256_init(&sha256_ctx);
    mbedtls_sha256_starts_ret(&sha256_ctx, false);

    while (len--) {
        if (BoxedBuffer::isInstance(*bufs)) {
            auto buf = (BoxedBuffer *)*bufs;
            mbedtls_sha256_update_ret(&sha256_ctx, buf->data, buf->length);
        }
        bufs++;
    }

    mbedtls_sha256_finish_ret(&sha256_ctx, output);

    return mkBuffer(output, sizeof(output));
}

} // namespace crypto

File: libs/core---esp32/crypto.ts
namespace crypto {
    /**
     * Compute cryptographic SHA256 hash of the concatenation of buffers. Returns 32-byte buffer.
     */
    //% promise shim=crypto::_sha256
    export declare function sha256(buffers: Buffer[]): Buffer;


    /**
     * Compute keyed-Hash Message Authentication Code as defined in RFC 2104.
     */
    export function sha256Hmac(key: Buffer, msg: Buffer) {
        const blockSize = 64
        if (key.length > blockSize) key = sha256([key])
        const paddedKey = Buffer.create(blockSize)
        paddedKey.write(0, key)
        for (let i = 0; i < blockSize; ++i) paddedKey[i] ^= 0x36
        const h0 = sha256([paddedKey, msg])
        for (let i = 0; i < blockSize; ++i) paddedKey[i] ^= (0x36 ^ 0x5c)
        return sha256([paddedKey, h0])
    }

} // namespace crypto